File: 394_Decode_String/394_1.java
class Solution {
    public String decodeString(String s) {
        Deque<Integer> numStack = new ArrayDeque<>();
        Deque<String> resStack = new ArrayDeque<>();
        int num = 0;
        StringBuilder res = new StringBuilder();
        for (char c : s.toCharArray()) {
            if (c >= '0' && c <= '9') {
                num = 10 * num + (c - '0');
            } else if (c == '[') {
                numStack.addLast(num);
                resStack.addLast(res.toString());
                num = 0;
                res = new StringBuilder();
            } else if (c == ']') {
                StringBuilder tmp = new StringBuilder();
                int curNum = numStack.removeLast();
                for (int i = 0; i < curNum; i++) {
                    tmp.append(res);
                }
                res = new StringBuilder(resStack.removeLast() + tmp);
            } else {
                res.append(c);
            }
        }
        return res.toString();
    }
}

File: 394_Decode_String/394_1.cpp
class Solution {
public:
    string decodeString(string s) {
        string res = "";
        int num = 0;
        stack<int> numStack;
        stack<string> resStack;
        for (char c : s) {
            if (c >= '0' && c <= '9') {
                num = num * 10 + (c - '0');
            } else if (c == '[') {
                numStack.push(num);
                resStack.push(res);
                num = 0;
                res.clear();
            } else if (c == ']') {
                string tmp = resStack.top();
                resStack.pop();
                int curNum = numStack.top();
                numStack.pop();
                for (int i = 0; i < curNum; i++) {
                    tmp += res;
                }
                res = tmp;
            } else {
                res.push_back(c);
            }
        }
        return res;
    }
};
